### PR TITLE
Have Hash#replace/dup/initialize_copy not rehash [Bug #16996]

### PR DESCRIPTION
**Pull request by marcandre**

### CODE REVIEW

**Comment by marcandre** [comment]
Follow up in #4273

File: hash.c [hunk 1]
```diff
@@ -2976,7 +2976,7 @@ rb_hash_replace(VALUE hash, VALUE hash2)
             RHASH_TBL_RAW(hash)->type = RHASH_ST_TABLE(hash2)->type;
         }
     }
-    rb_hash_foreach(hash2, rb_hash_rehash_i, (VALUE)hash);
+	hash_copy(hash, hash2);
```
Comment (nobu) [inline hash.c:2979]:
```suggestion
    hash_copy(hash, hash2);
```

File: test/ruby/test_hash.rb [hunk 2]
```diff
@@ -113,13 +113,18 @@ def test_self_initialize_copy
...
-    set2 = @cls[set1 => true]
+  def test_dup_will_not_rehash
+    x = []
+    set1 = @cls[x, true]
+    x << 42 # set1 need to be rehashed
+    set2 = set1.dup
+    set1.rehash
 
-    set1[set1] = true
+    assert_not_equal set1, set2
```
Comment (nobu) [inline test/ruby/test_hash.rb:123]: Same as https://github.com/ruby/ruby/pull/4273#discussion_r594872410

**Review by jeremyevans — COMMENTED** [review]
This looks good to me, after the changes @nobu recommended.  Hopefully this gets accepted at the developer meeting.

**Comment by marcandre** [comment]
> Hopefully this gets accepted at the developer meeting.

I don't think it is necessary to ask at the dev meeting. There has been only support in the past 9 months. It is spec that it is a mistake to mutate a key without calling `rehash` (and otherwise this has no impact). Also, two methods have changed in the past, without anyone asking (or even noticing).

**Comment by marcandre** [comment]
Tweaked with improved test. Now includes #4273

**Comment by jeremyevans** [comment]
Hash#dup/replace/initialize_copy not rehashing was approved at the development meeting: https://github.com/ruby/dev-meeting-log/blob/master/DevelopersMeeting20210317Japan.md#bug-16996-hash-should-avoid-doing-unnecessary-rehash-jeremyevans0

So it should be fine to merge this.

**Comment by marcandre** [comment]
Great, thanks!

**Pull request closed without merging**